### ISSUE TITLE
[BUG] Affliction & Demonology Mastery are rounded down

### ISSUE THREAD

**Issue by Tharre** [JamminL/cata-classic-bugs#1256]
#### 1.1. Description

Both affliction and demonology's respective masteries final percentages are rounded down instead of being continuous. So for example as affliction, with `1063` mastery rating or `13.93` total mastery points the total bonus from mastery will be rounded down from `0.227047` to `0.22`.

  #### 1.2. How to Reproduce

1. Calculate your total bonus from mastery, and adjust your gear such that you're well below the next percentage point.
2. Look at the damage of a (shadow) spell with no variance, for example drain soul
3. Adjust mastery slightly, for example with a gem, but make sure you're still below the next percentage point
4. Look at the damage of the spell from step 2. again, it's still the same

For demonology you can follow the same steps, except you need to do them during metamorphosis (be aware that it snapshots, so you need to cast it at least twice).

Same with demonology pet damage, by looking at the damage range of the pet in the character frame.

  #### 1.3  Source Material
#### 2.1. Description

The mastery bonus should be continuously affecting damage, not be rounded down.

  #### 2.2 Source Material

None yet. But this is clearly not intended given it does work as expected with the destruction mastery on the same class. Other classes masteries are also claimed to be continuous, though I've not tested that myself.

**Comment by ssabbyccatt** [comment]
This is true for balance druid as well

**Comment by NerdEgghead** [comment]
> This is true for balance druid as well (I meant to submit a bug report on this a long time ago, oops)

I believe the current list of bugged specs for Mastery flooring is: Affliction Warlock, Demonology Warlock, Arcane Mage, Fire Mage, Balance Druid, and Resto Druid. I'll go ahead and add Mage and Druid tags to this bug report.

**Comment by Zereges** [comment]
I wonder if someone can find source for this with how little difference this makes.

**Comment by ssabbyccatt** [comment]
> I wonder if someone can find source for this with how little difference this makes.

Source for it not rounding? Easiest source tbh is 4.3 Rawr or Sim-C, neither of which round. 
If you mean a source for how it impacts dmg/gameplay, it's not a significant impact on dmg (at least for boomie - I can't speak for other specs). It's a bit annoying for gearing though because we have to gear around haste breakpoints + mastery breakpoints right now (instead of just haste breakpoints).

**Comment by NerdEgghead** [comment]
The main motivation for fixing this wouldn't be because of some large numerical impact on spec performance or class balance, etc., but moreso that behavior like this is extremely unintuitive to players. Seeing your DPS not change at all when adding more of a stat whose stated purpose is to add a multiplier on your DPS ends up being pretty confusing. It also creates more of a reliance on computational tools (which we do have already) to min-max all your Reforges to precisely hit these artificial integer % breakpoints without wasting any stat budget. While _creating_ these types of optimization tools is something I personally enjoy doing, I dislike that the average player's gearing becomes dependent on them rather than being able to just intuitively add more of a good stat and expecting your numbers to go up as advertised.

**Comment by kenthklui** [comment]
Renaming the issue to better reflect it as a multi-spec issue

**Comment by Artemishowl** [comment]
Throwing in a ticket, thank you!

**Comment by TheBackstabi** [comment]
To add on here, this also applies to all three Rogue specializations. I suspect this is just the current functionality of Classic to ignore/truncate the final Mastery % value, like TBC expertise

This was a test done as Assassination; currently trying to hunt down a test that was done as Combat but I lost the link:
![image](https://github.com/user-attachments/assets/ab516e59-9a7f-474d-8120-79a4c5e3e9bf)

**Comment by Tharre** [comment]
> I suspect this is just the current functionality of Classic to ignore/truncate the final Mastery % value, like TBC expertise

No, some specs, like destro warlock, do not have this issue. It's very much a spec thing and not like TBC expertise.

**Comment by Tharre** [comment]
Seems to have been partially fixed, affliction mastery now works properly and apparently other specs as well (though I haven't tested those myself).

Demo mastery is still partially broken however, it works for metamorphosis but the pet damage portion is still rounded down.
